core/storage/keystore/keystore.hpp

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FILECOIN_CORE_STORAGE_KEYSTORE_HPP
+#define FILECOIN_CORE_STORAGE_KEYSTORE_HPP
+
+#include <boost/variant.hpp>
+#include <gsl/span>
+#include <libp2p/crypto/secp256k1_provider.hpp>
+#include <libp2p/crypto/secp256k1_types.hpp>
+#include <libp2p/outcome/outcome.hpp>
+
+#include "crypto/bls_provider/bls_provider.hpp"
+#include "crypto/bls_provider/bls_types.hpp"
+#include "primitives/address.hpp"
+#include "primitives/address/address_verifier.hpp"
+#include "storage/keystore/keystore_error.hpp"
+
+namespace fc::storage::keystore {
+
+  using crypto::bls::BlsProvider;
+  using libp2p::crypto::secp256k1::Secp256k1Provider;
+  using primitives::Address;
+  using BlsKeyPair = fc::crypto::bls::KeyPair;
+  using BlsPrivateKey = fc::crypto::bls::PrivateKey;
+  using BlsSignature = fc::crypto::bls::Signature;
+  using Secp256k1KeyPair = libp2p::crypto::secp256k1::KeyPair;
+  using Secp256k1PrivateKey = libp2p::crypto::secp256k1::PrivateKey;
+  using Secp256k1Signature = libp2p::crypto::secp256k1::Signature;
+  using fc::primitives::address::AddressVerifier;
+
+  /**
+   * An interface to a facility to store and use cryptographic keys
+   */
+  class KeyStore {
+   public:
+    using TPrivateKey = boost::variant<BlsPrivateKey, Secp256k1PrivateKey>;
+    using TSignature = boost::variant<BlsSignature, Secp256k1Signature>;
+
+    KeyStore(std::shared_ptr<BlsProvider> blsProvider,
+             std::shared_ptr<Secp256k1Provider> secp256K1Provider,
+             std::shared_ptr<AddressVerifier> addressVerifier);
+
+    virtual ~KeyStore() = default;
+
+    /**
+     * @brief Whether or not key exists in the Keystore
+     * @param address key identifier
+     */
+    virtual outcome::result<bool> Has(const Address &address) noexcept = 0;
+
+    /**
+     * @brief stores a key in the Keystore
+     * @param address - key idnetifier
+     * @param key
+     */
+    virtual outcome::result<void> Put(Address address,
```

Comment:
Why return a variant here, not just KeyStoreError?

Reply:
It's common practice for our projects to return outcome::result to handle error-or-result.
For consistency methods that don't return value but may return error should return outcome::result<void>.

Reply:
To use wrappers OUTCOME_TRY....